### PR TITLE
Don't poll for disconnects in BaseHTTPMiddleware via StreamingResponse

### LINKED ISSUE #2516: `RuntimeError("No response returned")` in `BaseHTTPMiddleware`
Previously, a "No response returned"-error has been discussed in this discussion: https://github.com/encode/starlette/discussions/1527, and was supposedly fixed in this PR: https://github.com/encode/starlette/pull/1715

However, the bug is still present. It's a race condition, so quite tough to reproduce, but the following reproducer works.

Reproducer
--

Use anyio 4.2.0 It does happen also with anyio 3.0.0, but it happens much more regularly on 4.2.

First, apply the following patch to the Starlette codebase, so that we have a breakpoint in case it happens:
```diff
--- a/starlette/middleware/base.py
+++ b/starlette/middleware/base.py
@@ -134,7 +134,8 @@ class BaseHTTPMiddleware:
             async def send_no_error(message: Message) -> None:
                 try:
@@ -163,6 +164,7 @@ class BaseHTTPMiddleware:
             except anyio.EndOfStream:
                 if app_exc is not None:
                     raise app_exc
+                breakpoint()
                 raise RuntimeError("No response returned.")

             assert message["type"] == "http.response.start"
```

Then run this code:

```python
import asyncio
from hypercorn.asyncio import serve
from hypercorn.config import Config

from starlette.applications import Starlette
from starlette.responses import JSONResponse
from starlette.routing import Route
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.requests import Request
from starlette.responses import Response
from starlette.middleware.base import RequestResponseEndpoint
from starlette.middleware import Middleware


class DummyMiddleware(BaseHTTPMiddleware):
    async def dispatch(
        self, request: Request, call_next: RequestResponseEndpoint
    ) -> Response:
        return await call_next(request)


async def homepage(request):
    return JSONResponse({})


app = Starlette(
    routes=[
        Route("/", homepage),
    ],
    middleware=[Middleware(DummyMiddleware) for i in range(100)],  # Yes, 100x the same middleware.
)

config = Config.from_mapping({})
config.bind = ["127.0.0.1:8000"]
asyncio.run(serve(app, config))
```

Open either Firefox or Chrome, navigate to http://localhost:8000 and hold down command-r (refresh) to repeatedly have the browser refresh the page. After a couple of seconds, we'll enter the breakpoint.

(Feel free to turn this issue into a discussion if that's more appropriate. I'm also looking for a fix but have a hard time understanding `BaseHTTPMiddleware`.)

edit: I updated the above snippet and added the missing imports.

This PR contains a possible fix: https://github.com/encode/starlette/pull/2519

> [!IMPORTANT]
> - We're using [Polar.sh](https://polar.sh/encode) so you can upvote and help fund this issue.
> - We receive the funding once the issue is completed & confirmed by you.
> - Thank you in advance for helping prioritize & fund our backlog.

<a href="https://polar.sh/encode/starlette/issues/2516">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/encode/starlette/issues/2516/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by adriangb**
Fixes #2516

### CODE REVIEW

File: tests/middleware/test_base.py [hunk 1]
```diff
@@ -1035,3 +1036,86 @@ async def endpoint(request: Request) -> Response:
...
+            self.events.append(f"{self.version}:STARTED")
+            try:
+                res = await call_next(request)
+                self.events.append(f"{self.version}:COMPLETED")
+                return res
+            except Exception as exc:
+                self.events.append(f"{self.version}:FAILED:{type(exc)}")
+                return Response(b"")
+
+    async def sleepy(request: Request):
```
Comment (mikkelduif) [inline tests/middleware/test_base.py]:
```suggestion
    async def sleepy(request: Request) -> Response:
```

**Comment by mikkelduif** [comment]
I can confirm that this PR fixes the example I have reproduced in https://github.com/encode/starlette/issues/2516#issuecomment-1976400404

**Comment by mikkelduif** [comment]
> I do still recommend you avoid `BaseHTTPMiddleware`

Maybe we should consider to add a section to: https://www.starlette.io/middleware/#basehttpmiddleware emphasising this.
I do not know the exact technicalities for this, however, it might help future users to avoid it.

**Comment by adriangb** [comment]
Seems like the issue now is coverage _in tests_ (all of the tests still pass). There are several pieces of _test_ code that are not not being hit but I'm not sure if they were essential or just an implementation detail, some more careful evaluation is needed.

**Comment by avihaiyosef** [comment]
@adriangb, what is the status of this PR?

**Comment by dlahyani** [comment]
Hi @adriangb , thanks for this fix.
Any expectation on when this is going to land in a new release of starlette?

**Comment by Kludex** [comment]
This PR needs someone motivated to check why the coverage dropped.

Anyone that uses BaseHTTPMiddleware in their company, and wants to spend some engineers hours working on this, is welcome.

Otherwise, you can try to sponsor @adriangb as an appreciation gesture for his efforts here, and maybe motivate him to continue the work here.

**Comment by isidentical** [comment]
@Kludex thanks for the reminder to sponsor, @adriangb has been doing amazing work and we just hit an issue that might be related to this PR so we just started sponsoring (albeit modestly) thru github!!! Would highly recommend anyone else to do the same.

**Comment by adriangb** [comment]
@Kludex I looked into the coverage.

Basically before we were using `StreamingResponse` which calls `receive()` to check for disconnection.
Since I changed this to use a similar implementation but one that does not check for disconnection `receive()` is no longer called.

For example, if I check out master and apply the changes to `test_base.py` I get the following traceback:

```
tests/middleware/test_base.py:776: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
starlette/middleware/base.py:189: in __call__
    with collapse_excgroups():
../../.pyenv/versions/3.12.2/lib/python3.12/contextlib.py:158: in __exit__
    self.gen.throw(value)
starlette/_utils.py:89: in collapse_excgroups
    raise exc
starlette/responses.py:265: in __call__
    await wrap(partial(self.listen_for_disconnect, receive))
starlette/responses.py:261: in wrap
    await func()
starlette/responses.py:238: in listen_for_disconnect
    message = await receive()
starlette/middleware/base.py:54: in wrapped_receive
    msg = await self.receive()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

    async def rcv() -> AsyncGenerator[Message, None]:
        yield {"type": "http.request", "body": b"1", "more_body": True}
        yield {"type": "http.request", "body": b"2", "more_body": True}
        yield {"type": "http.request", "body": b"3"}
>       raise AssertionError(  # pragma: no cover
            "Should not be called, no need to poll for disconnect"
        )
E       AssertionError: Should not be called, no need to poll for disconnect
```

**Comment by adriangb** [comment]
> @Kludex thanks for the reminder to sponsor, @adriangb has been doing amazing work and we just hit an issue that might be related to this PR so we just started sponsoring (albeit modestly) thru github!!! Would highly recommend anyone else to do the same.

Thank you for your sponsorship!

**Comment by hakanutku** [comment]
@Kludex Could you please review this PR when you get a chance? I appreciate your time!

**Comment by Kludex** [comment]
I'm a bit tired of people asking me stuff.

I'm very dedicated to the projects I maintain, which means people just need to wait a bit. I don't like people pressuring me to review stuff.

You want to motivate me to check this PR? Feel free to sponsor me. Otherwise, wait for me to voluntarily come here and review it.

Too many issues regarding BaseHTTPMiddleware, and I need to check previous work and make sure this is fine. It takes time.

**Review by Kludex — APPROVED** [review]

File: starlette/middleware/base.py [hunk 2]
```diff
@@ -198,20 +198,33 @@ async def dispatch(
...
-class _StreamingResponse(StreamingResponse):
+class _StreamingResponse(Response):
     def __init__(
         self,
-        content: ContentStream,
+        content: AsyncContentStream,
         status_code: int = 200,
         headers: typing.Mapping[str, str] | None = None,
         media_type: str | None = None,
-        background: BackgroundTask | None = None,
```
Comment (dzhulgakov) [inline starlette/middleware/base.py:208]: is the removal of `background` here intentional? We use `background` for both actual background tasks but also to detect when a streaming request was cancelled by the client (`request.is_disconnected` had some minor overhead).

Comment (dmitry-mli) [reply starlette/middleware/base.py:208]: Cutting a PR for this

Comment (dmitry-mli) [reply starlette/middleware/base.py:208]: https://github.com/encode/starlette/pull/2687

**Pull request merged**